### PR TITLE
Add CubebDeviceCollectionManager to handle mapping between CubebServer and shared cubeb context.

### PR DESCRIPTION
**Pull request by kinetiknz**
The previous implementation of cubeb_register_device_collection_changed in https://github.com/djg/audioipc-2/pull/62 applied the registration directly to the cubeb context.  CubebServer logically represents a cubeb context, but the underlying cubeb context is actually shared (via TLS) to multiple CubebServers (with one CubebServer per attached client).

The fix requires introducing a layer to manage device collection changed callbacks for multiple CubebServers using a single cubeb context.  This PR introduces a CubebDeviceCollectionManager that rides along with the shared TLS cubeb context and maps multiple CubebServer registrations to the shared cubeb context.

r? @ChunMinChang please

### CODE REVIEW

File: server/src/server.rs [hunk 1]
```diff
@@ -36,26 +37,161 @@ fn error(error: cubeb::Error) -> ClientMessage {
...
+    input_registered: bool,
+    output_registered: bool,
+}
+
+impl CubebDeviceCollectionManager {
+    fn new() -> CubebDeviceCollectionManager {
+        CubebDeviceCollectionManager {
+            servers: Vec::new(),
+            input_registered: false,
+            output_registered: false,
```
Comment (ChunMinChang) [inline server/src/server.rs]: Have you considered using `DeviceType` for registered scope?

File: server/src/server.rs [hunk 2]
```diff
@@ -36,26 +37,161 @@ fn error(error: cubeb::Error) -> ClientMessage {
...
+        }
+    }
+
+    fn register(&mut self, context: &cubeb::Context, server: &Rc<RefCell<CubebServerCallbacks>>) {
+        if self
```
Comment (ChunMinChang) [inline server/src/server.rs]: Check the `devtype` of the `CubebServerCallbacks`?

File: server/src/server.rs [hunk 3]
```diff
@@ -36,26 +37,161 @@ fn error(error: cubeb::Error) -> ClientMessage {
...
+            .is_none()
+        {
+            self.servers.push(server.clone());
+        }
+        self.update(context);
+    }
+
+    fn unregister(&mut self, context: &cubeb::Context, server: &Rc<RefCell<CubebServerCallbacks>>) {
+        self.servers
+            .retain(|s| !(Rc::ptr_eq(&s, server) && s.borrow().devtype.is_empty()));
```
Comment (ChunMinChang) [inline server/src/server.rs]: Why using `retain` instead of `remove_item`?

File: server/src/server.rs [hunk 4]
```diff
@@ -36,26 +37,161 @@ fn error(error: cubeb::Error) -> ClientMessage {
...
+                    .register_device_collection_changed(
+                        cubeb::DeviceType::OUTPUT,
+                        None,
+                        std::ptr::null_mut(),
+                    )
+                    .expect("output devcol unregister failed");
+                self.output_registered = false;
+            }
+        }
+    }
```
Comment (ChunMinChang) [inline server/src/server.rs]: Is it possible to merge the duplicated code in `internal_unregister` and `internal_register`?

File: server/src/server.rs [hunk 5]
```diff
@@ -413,35 +561,14 @@ impl CubebServer {
...
-                    cb,
-                    user_ptr,
-                )?;
-            }
+        if enable {
+            cbs.borrow_mut().devtype.insert(devtype);
+            manager.register(context, cbs);
+        } else {
+            cbs.borrow_mut().devtype.remove(devtype);
+            manager.unregister(context, cbs);
```
Comment (ChunMinChang) [inline server/src/server.rs]: Looks better now!

File: server/src/server.rs (hunk 3 above)
Comment (kinetiknz) [reply server/src/server.rs]: remove_item is unstable (nightly only).

**Comment by kinetiknz** [comment]
> My comments in the previous round are lost, so I post them again. Sorry about that.

Sorry about that! I rebased the patch on master after the reformatting landed, I wasn't sure if you'd started reviewing this yet so assumed it might be safe - I should've checked with you.

**Pull request merged**